docs/design/chart_format.md

```diff
@@ -0,0 +1,664 @@
+
+
+A generator reads through the files in `templates/` and looks for special file headers that indicate what generation tool needs to be run (if any) on a file. One such implementation of a generation tool is a template renderer. In any one pass through the `templates/` directory, more than one generation tool may be invoked.
+
+
+#### The Generator Header
+A file that requires a generation tool must declare this on the first line of the file using one of the following syntaxes:
+
+```
+//helm:generate TOOL [ARGS]\n
+#helm:generate TOOL [ARGS]\n
+/*helm:generate TOOL [ARGS]*/
+```
+
+`TOOL` corresponds to a registered tool name (see [this proposal](https://github.com/kubernetes/kubernetes/pull/18215#issuecomment-168885511)). `[ARGS]` are zero or more space-separated strings that will be passed unmodified to the tool. The first two formats require a terminating line feed. The third format requires a terminating `*/` sequence.
+
+
+**Example:**
+
+```
+#helm:generate dm_template
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+
+In the example above, the `#helm:generate dm_template` indicates that the remainder of the file contents should be processed using whatever handler is registered for handling `dm_template`.
+
+
+Generator tools may take arguments. No format is specified for arguments. For example, one might use UNIX-like flags.
+
+
+```
+#helm:generate dm_template --ignore-timezone -s
+{% set PROPERTIES = properties or {} %}
+{% set WORKERS = PROPERTIES['workers'] or 2 %}
+resources:
+- name: spark-master
+ type: ReplicationController
+ properties:
+   kind: ReplicationController
+   apiVersion: v1
+   metadata:
+     name: spark-master-controller
+   spec:
+…
+```
+
+Any arguments are passed into the generator unaltered. Generators may ignore arguments that they do not support or understand.
+
+
+A generator tool _must_ not overwrite any files (including itself) present in the chart’s `templates/` directory. It must not alter the `README.md`, the `LICENSE`, the `Chart.yaml`, or the contents of the `hooks/` directory. These restrictions maintain the reproducibility of the build over subsequent generator runs.
+
+
+One common form of generator tool is a template renderer. Such a tool takes a template (and possibly other supporting files) as input, processes the template, and outputs one or more generated files or data streams. The location of output is not specified by this proposal (though it is subject to the caveats above).
+
+
+An implementation of this specification may support any number of template engines. However, two are explicitly defined: Python/Jinja templates (formerly DM Templates), and simple Go templates (formerly Helm Templates).
```

Comment:
We should make an effort to include Jsonnet early on, which is worth mentioning here.

/cc @sparkprime